<robot name="rc_backhoe" xmlns:xacro="http://ros.org/wiki/xacro">
  <xacro:arg name="robot_name" default="/rc_backhoe" />
  <xacro:property name="robot_name" value="$(arg robot_name)" />

  <material name="red">
    <color rgba="1.0 0.0 0.0 1.0"/>
  </material>
  <material name="gray">
    <color rgba="0.4 0.4 0.6 1.0"/>
  </material>
  <material name="light_gray">
    <color rgba="0.8 0.8 0.8 1.0"/>
  </material>

  <xacro:macro name="arm4d_macro" params="l1 l2 l3 l4 l5">
    <link name="base_link"/>

    <joint name="body_joint" type="fixed">
      <parent link="base_link"/>
      <child link="body_link"/>
    </joint>
    <link name="body_link">
      <visual>
        <origin rpy="0 0 0" xyz="0 0 ${l1/2}"/>
        <geometry>
          <box size="0.25 0.25 ${l1}"/>
        </geometry>
        <material name="gray"/>
      </visual>
      <collision>
        <origin rpy="0 0 0" xyz="0 0 ${l1/2}"/>
        <geometry>
          <box size="0.25 0.25 ${l1}"/>
        </geometry>
      </collision>
      <inertial>
        <origin rpy="0 0 0" xyz="0 0 ${l1/2}"/>
        <mass value="50"/>
        <inertia ixx="10.0" ixy="0.0" ixz="0.0" iyy="10.0" iyz="0.0" izz="10.0"/>
      </inertial>
    </link>

    <joint name="swing_joint" type="revolute">
      <parent link="body_link"/>
      <child link="swing_link"/>
      <origin rpy="0 0 0" xyz="0 0 ${l1}"/>
      <axis xyz="0 0 1"/>
      <limit effort="30" lower="${radians(-180)}" upper="${radians(180)}" velocity="0.5"/>
      <dynamics damping="1.0" friction="20.0" />
    </joint>
    <link name="swing_link">
      <visual>
        <origin rpy="0 0 0" xyz="0 0 ${l2/2}"/>
        <geometry>
          <box size="0.05 0.05 ${l2}"/>
        </geometry>
        <material name="light_gray"/>
      </visual>
      <collision>
        <origin rpy="0 0 0" xyz="0 0 ${l2/2}"/>
        <geometry>
          <box size="0.05 0.05 ${l2}"/>
        </geometry>
      </collision>
      <inertial>
        <origin rpy="0 0 0" xyz="0 0 ${l2/2}"/>
        <mass value="0.3"/>
        <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
      </inertial>
    </link>

    <joint name="boom_joint" type="revolute">
      <parent link="swing_link"/>
      <child link="boom_link"/>
      <origin rpy="0 ${radians(30)} 0" xyz="0 0 ${l2}"/>
      <axis xyz="0 1 0"/>
      <limit effort="30" lower="${radians(0)}" upper="${radians(90)}" velocity="0.5"/>
      <dynamics damping="1.0" friction="20.0" />
    </joint>
    <link name="boom_link">
      <visual>
        <origin rpy="0 0 0" xyz="0 0 ${l3/2}"/>
        <geometry>
          <box size="0.05 0.05 ${l3}"/>
        </geometry>
        <material name="gray"/>
      </visual>
      <collision>
        <origin rpy="0 0 0" xyz="0 0 ${l3/2}"/>
        <geometry>
          <box size="0.05 0.05 ${l3}"/>
        </geometry>
      </collision>
      <inertial>
        <origin rpy="0 0 0" xyz="0 0 ${l3/2}"/>
        <mass value="0.3"/>
        <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
      </inertial>
    </link>

    <joint name="arm_joint" type="revolute">
      <parent link="boom_link"/>
      <child link="arm_link"/>
      <origin rpy="0 ${radians(30)} 0" xyz="0 0 ${l3}"/>
      <axis xyz="0 1 0"/>
      <limit effort="30" lower="${radians(0)}" upper="${radians(90)}" velocity="0.5"/>
      <dynamics damping="1.0" friction="20.0" />
    </joint>
    <link name="arm_link">
      <visual>
        <origin rpy="0 0 0" xyz="0 0 ${l4/2}"/>
        <geometry>
          <box size="0.05 0.05 ${l4}"/>
        </geometry>
        <material name="light_gray"/>
      </visual>
      <collision>
        <origin rpy="0 0 0" xyz="0 0 ${l4/2}"/>
        <geometry>
          <box size="0.05 0.05 ${l4}"/>
        </geometry>
      </collision>
      <inertial>
        <origin rpy="0 0 0" xyz="0 0 ${l4/2}"/>
        <mass value="0.3"/>
        <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
      </inertial>
    </link>

    <joint name="bucket_joint" type="revolute">
      <parent link="arm_link"/>
      <child link="bucket_link"/>
      <origin rpy="0 ${radians(0)} 0" xyz="0 0 ${l4}"/>
      <axis xyz="0 1 0"/>
      <limit effort="30" lower="${radians(0)}" upper="${radians(90)}" velocity="0.5"/>
      <dynamics damping="1.0" friction="20.0" />
    </joint>
    <link name="bucket_link">
      <visual>
        <origin rpy="0 0 0" xyz="0 0 ${l5/2}"/>
        <geometry>
          <box size="0.02 0.02 ${l5}"/>
        </geometry>
        <material name="gray"/>
      </visual>
      <collision>
        <origin rpy="0 0 0" xyz="0 0 ${l5/2}"/>
        <geometry>
          <box size="0.02 0.02 ${l5}"/>
        </geometry>
      </collision>
      <inertial>
        <origin rpy="0 0 0" xyz="0 0 ${l5/2}"/>
        <mass value="0.3"/>
        <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
      </inertial>
    </link>

    <joint name="target_joint" type="fixed">
      <parent link="bucket_link"/>
      <child link="target_link"/>
      <origin rpy="0 0 0" xyz="0 0 ${l5}"/>
    </joint>
    <link name="target_link">
      <visual>
        <geometry>
          <sphere radius="0.02"/>
        </geometry>
        <material name="red"/>
      </visual>
      <collision>
        <geometry>
          <sphere radius="0.02"/>
        </geometry>
      </collision>
      <inertial>
        <mass value="0.3"/>
        <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
      </inertial>
    </link>

    <transmission name="swing_trans">
      <type>transmission_interface/SimpleTransmission</type>
        <joint name="swing_joint">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      </joint>
      <actuator name="swing_motor">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
    </transmission>
    <transmission name="boom_trans">
      <type>transmission_interface/SimpleTransmission</type>
        <joint name="boom_joint">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      </joint>
      <actuator name="boom_motor">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
    </transmission>
    <transmission name="arm_trans">
      <type>transmission_interface/SimpleTransmission</type>
        <joint name="arm_joint">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      </joint>
      <actuator name="arm_motor">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
    </transmission>
    <transmission name="bucket_trans">
      <type>transmission_interface/SimpleTransmission</type>
        <joint name="bucket_joint">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      </joint>
      <actuator name="bucket_motor">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
    </transmission>
    <gazebo>
      <plugin name="rc_backhoe_retrofit_model" filename="librc_backhoe_retrofit_model.so">
      </plugin>
      <!-- <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so"> -->
      <!-- <plugin name="gazebo_ros_control" filename="librc_backhoe_hw_sim.so"> -->
        <!-- <robotNamespace>/</robotNamespace> -->
        <!-- <robotSimType>gazebo_ros_control/DefaultRobotHWSim</robotSimType> -->
        <!-- <robotSimType>gazebo_ros_control/RcBackhoeHWSim</robotSimType> -->
        <!-- <legacyModeNS>true</legacyModeNS> -->
      <!-- </plugin> -->
    </gazebo>

    <link name="swing_imu_link" />
    <joint name="swing_imu_joint" type="fixed">
      <parent link="swing_link" />
      <child link="swing_imu_link" />
      <origin rpy="0 0 0" xyz="0 0 0"/>
    </joint>  
    <gazebo reference="swing_imu_link">
    <gravity>true</gravity>
    <sensor name="swing_imu_sensor" type="imu">
      <always_on>true</always_on>
      <update_rate>100</update_rate>
      <visualize>true</visualize>
      <topic>__default_topic__</topic>
      <plugin filename="libgazebo_ros_imu_sensor.so" name="imu_plugin">
        <topicName>${robot_name}/swing/imu/data_raw</topicName>
        <bodyName>swing_imu_link</bodyName>
        <updateRateHZ>10.0</updateRateHZ>
        <gaussianNoise>0.0</gaussianNoise>
        <xyzOffset>0 0 0</xyzOffset>
        <rpyOffset>0 0 0</rpyOffset>
        <frameName>swing_imu_link</frameName>
        <initialOrientationAsReference>false</initialOrientationAsReference>
      </plugin>
      <pose>0 0 0 0 0 0</pose>
    </sensor>
    </gazebo>

    <link name="boom_imu_link" />
    <joint name="boom_imu_joint" type="fixed">
      <parent link="boom_link" />
      <child link="boom_imu_link" />
      <origin rpy="${radians(-90)} ${radians(20)} 0" xyz="0 0 ${l3/2}"/>
    </joint>  
    <gazebo reference="boom_imu_link">
    <gravity>true</gravity>
    <sensor name="boom_imu_sensor" type="imu">
      <always_on>true</always_on>
      <update_rate>100</update_rate>
      <visualize>true</visualize>
      <topic>__default_topic__</topic>
      <plugin filename="libgazebo_ros_imu_sensor.so" name="imu_plugin">
        <topicName>${robot_name}/boom/imu/data_raw</topicName>
        <bodyName>boom_imu_link</bodyName>
        <updateRateHZ>10.0</updateRateHZ>
        <gaussianNoise>0.0</gaussianNoise>
        <xyzOffset>0 0 0</xyzOffset>
        <rpyOffset>0 0 0</rpyOffset>
        <frameName>boom_imu_link</frameName>
        <initialOrientationAsReference>false</initialOrientationAsReference>
      </plugin>
      <pose>0 0 0 0 0 0</pose>
    </sensor>
    </gazebo>

    <link name="arm_imu_link" />
    <joint name="arm_imu_joint" type="fixed">
      <parent link="arm_link" />
      <child link="arm_imu_link" />
      <origin rpy="${radians(-90)} ${radians(-71)} 0" xyz="0 0 ${l4/2}"/>
    </joint>  
    <gazebo reference="arm_imu_link">
    <gravity>true</gravity>
    <sensor name="arm_imu_sensor" type="imu">
      <always_on>true</always_on>
      <update_rate>100</update_rate>
      <visualize>true</visualize>
      <topic>__default_topic__</topic>
      <plugin filename="libgazebo_ros_imu_sensor.so" name="imu_plugin">
        <topicName>${robot_name}/arm/imu/data_raw</topicName>
        <bodyName>arm_imu_link</bodyName>
        <updateRateHZ>10.0</updateRateHZ>
        <gaussianNoise>0.0</gaussianNoise>
        <xyzOffset>0 0 0</xyzOffset>
        <rpyOffset>0 0 0</rpyOffset>
        <frameName>arm_imu_link</frameName>
        <initialOrientationAsReference>false</initialOrientationAsReference>
      </plugin>
      <pose>0 0 0 0 0 0</pose>
    </sensor>
    </gazebo>

    <link name="bucket_imu_link" />
    <joint name="bucket_imu_joint" type="fixed">
      <parent link="bucket_link" />
      <child link="bucket_imu_link" />
      <origin rpy="${radians(-90)} ${radians(-80)} 0" xyz="0 0 ${l5/4}"/>
    </joint>  
    <gazebo reference="bucket_imu_link">
    <gravity>true</gravity>
    <sensor name="bucket_imu_sensor" type="imu">
      <always_on>true</always_on>
      <update_rate>100</update_rate>
      <visualize>true</visualize>
      <topic>__default_topic__</topic>
      <plugin filename="libgazebo_ros_imu_sensor.so" name="imu_plugin">
        <topicName>${robot_name}/bucket/imu/data_raw</topicName>
        <bodyName>bucket_imu_link</bodyName>
        <updateRateHZ>10.0</updateRateHZ>
        <gaussianNoise>0.0</gaussianNoise>
        <xyzOffset>0 0 0</xyzOffset>
        <rpyOffset>0 0 0</rpyOffset>
        <frameName>bucket_imu_link</frameName>
        <initialOrientationAsReference>false</initialOrientationAsReference>
      </plugin>
      <pose>0 0 0 0 0 0</pose>
    </sensor>
    </gazebo>
  </xacro:macro>  

  <xacro:arm4d_macro l1="0.05" l2="0.05" l3="0.25" l4="0.13" l5="0.09"/>
</robot>